### PR TITLE
fix: reorder fake tool call messages at tail to correct timing

### PR DESCRIPTION
**Pull request by Rail1bc**
### Motivation / 动机

通过向 `req.contexts` 尾部注入 `assistant(tool_calls) → tool(result)` 消息对，可以**强制** LLM 认为自己已经调用了某个工具并拿到了结果，从而：

- 跳过 LLM 自身的工具调用决策，**节省 token**
- **降低延迟**（不需要等 LLM 生成工具调用）
- **增加确定性**（固定或规则触发，不受模型输出波动影响，增强确定性）

这是一种通用手法，已在 LivingMemory 插件中长期记忆注入中使用，未来可能被更多插件用于各种确定性工具调用场景。

但在 `OpenAI_Provider._prepare_chat_payload()` 中，处理顺序是：

1. `deepcopy(contexts)` —— 此时尾部已含伪造工具调用对
2. `append(new_record)` —— 追加当前用户消息

最终发给 LLM API 的 messages 序列为：

```
... assistant(tool_calls=...), tool(result), user("...")
```

LLM 视角下，assistant 在用户提问之前就"预知"了查询内容。正确顺序应为：

```
... user("..."), assistant(tool_calls=...), tool(result)
```

该问题影响所有使用该手法的插件，以及所有走 `OpenAI_Provider` 的 LLM 路径

  Refs: #9450

  ### Modifications / 改动点

  **`astrbot/core/provider/sources/openai_source.py`**

  在 `_sanitize_assistant_messages()` 末尾新增检测与重排逻辑：

  1. 从尾部弹出 `user` 消息
  2. 用 while 循环收集所有 `assistant(tool_calls) + tool` 成对消息（验证 `tool_call_id` 匹配）
  3. 重排为 `user → assistant₁, tool₁ → ... → assistant_N, tool_N`

  支持多插件各自注入多轮伪造对的场景。

  - 不依赖具体插件实现，通用模式检测
  - 不改动 `req` 对象，不破坏其他 `on_llm_request` handler
  - 覆盖 `_query` / `_query_streaming` 全部路径
  - 真实工具调用场景下 `tool` 后不会直接跟 `user`，不会误杀

  此为轻量妥协修复。未来若实现专用的上下文操作钩子或伪造工具调用钩子，可考虑移除此方法。

  - [x] This is NOT a breaking change. / 这不是一个破坏性变更。

  ### Test Results / 测试结果

  逻辑验证（手动）：
  - 无伪造对：不触发，不变
  - 单对：`..., asst(tc), tool, user` → `..., user, asst(tc), tool`
  - 多对：`..., asst₁, tool₁, asst₂, tool₂, user` → `..., user, asst₁, tool₁, asst₂, tool₂`
  - 真实工具调用：`..., asst(tc), tool, asst(content), user` → 不变，不误杀

  ---

  ### Checklist / 检查清单

  - [ ] 😊 新功能已通过 Issue 与作者讨论
  - [x] 👀 我的更改经过了充分测试，测试结果已在上方提供
  - [x] 🤓 无新依赖引入
  - [x] 😮 无恶意代码

## Summary by Sourcery

Ensure fake tool call assistant/tool/user message sequences are reordered so user messages precede injected tool call pairs, while preserving real tool call ordering and markers across all compatible providers and serialization paths.

Bug Fixes:
- Correct message ordering when plugins inject fake assistant(tool_calls) and tool messages at the tail so the LLM sees user queries before tool call pairs.
- Prevent crashes and misordering when message lists contain non-dict entries or incomplete/mismatched tool call IDs.
- Avoid reordering and protect timing of real tool calls by marking them and excluding marked pairs from fake-tool-call reordering.
- Ensure internal real-tool-call markers are stripped from outbound payloads so provider APIs never receive internal flags.

Enhancements:
- Share a reusable fake tool call fixture across provider tests to keep scenarios consistent.
- Propagate and persist real-tool-call markers through Message serialization, ToolCallsResult conversion, modalities sanitization, and checkpoint roundtrips to support robust reordering logic.
- Apply the fake-tool-call reorder logic to OpenAI, Anthropic, Gemini, and OpenAI Responses providers, including both streaming and non-streaming code paths.
- Add comprehensive tests across providers to validate reordering behavior, marker preservation, and mixed real/fake tool call scenarios.

Tests:
- Extend unit tests for OpenAI, Anthropic, Gemini, and OpenAI Responses providers to cover fake tool call reordering, real tool call protection, marker stripping, and edge cases such as mixed pairs and malformed messages.

### CODE REVIEW

File: astrbot/core/provider/sources/openai_source.py [hunk 1]
```diff
@@ -528,6 +528,58 @@ def _is_empty(content: Any) -> bool:
             )
         payloads["messages"] = final
 
+        ProviderOpenAIOfficial._reorder_tailing_tool_call_user(payloads)
+
+    @staticmethod
+    def _reorder_tailing_tool_call_user(payloads: dict) -> None:
```
[the remark this answers is not in the document]
Comment (Rail1bc) [reply astrbot/core/provider/sources/openai_source.py]: 我认为这样修改几乎没有任何收益

**Review by xiaoyuyu6420 — COMMENTED** [review]
## Review

Clean fix for #9450. The approach of popping the trailing `user` message and re-inserting it before the `assistant(tool_calls) → tool` pairs is correct and minimal.

### Strengths

1. **Validates `tool_call_id` matching** — the code checks `tool_msg.get("tool_call_id") not in tc_ids` before treating a pair as a tool-call pair. This prevents misidentifying unrelated `assistant → tool` sequences.
2. **Reverses `pairs` before re-insertion** — correctly handles the LIFO collection order.
3. **Handles the no-op case** — when no fake pairs are found, the `user` message is put back in place.

### Issues

**1. No tests**

This is a non-trivial message reordering algorithm with several branches (no pairs, one pair, multiple pairs, mismatched tool_call_id). There should be at least one test per branch:
- Trailing user after a single fake tool call pair
- Trailing user after multiple fake tool call pairs
- No fake pairs (message order unchanged)
- assistant → tool without tool_calls (should stop collecting)

**2. Only handles the OpenAI official provider**

This fix is in `ProviderOpenAIOfficial`, but the issue (#9450) says the problem occurs in `_prepare_chat_payload`. If other providers (e.g. OpenAI-compatible providers) have the same `_prepare_chat_payload` pattern, they'll have the same bug. Worth checking if this needs to be applied more broadly.

**3. Edge case: what if the user message itself contains tool-call-related content?**

If a plugin injects a fake tool call AND the user message has role `"user"` but also contains function-related content, the reordering still treats it as a plain user message. This is probably correct (OpenAI expects user messages after tool results), but worth a comment.

**4. Performance: O(n) pop from the end in a while loop**

Each `messages.pop()` is O(1) from the end, and the while loop processes at most `len(messages)/2` iterations, so this is fine. Just noting that the list is mutated in place.

### Summary

Solid algorithm, correct logic. Main gap is missing tests — a message reordering function with 4+ branches should have test coverage before merging.

**Comment by Rail1bc** [comment]
修复伪造工具调用（fake tool call）消息对在 OpenAI / Anthropic / Gemini 格式 provider 中的时序错乱问题。

修复从openai扩展到OpenAI / Anthropic / Gemini 格式
覆盖了测试

### Modifications / 改动点




- 在 `astrbot/core/provider/provider.py` 新增共享函数 `reorder_tailing_tool_call_user(messages: list)`：从尾部弹出 user 消息，持续收集 `assistant(tool_calls) + tool` 成对消息（校验 `tool_call_id` 匹配），重排为 `user → assistant → tool`。
- 在 `astrbot/core/provider/__init__.py` 中 re-export 该函数。
- `openai_source.py`：删除原 static method `_reorder_tailing_tool_call_user`，改用共享函数（`_query` 与 `_query_streaming` 两条路径均覆盖）。
- `anthropic_source.py`：在 `text_chat` / `text_chat_stream` 调用 `_prepare_payload` 之前，对 OpenAI-format 的 `context_query` 应用同一重排函数。
- `gemini_source.py`：在 `text_chat` / `text_chat_stream` 组装 `context_query` 之后（`_prepare_conversation` 转换之前）应用同一重排函数。Gemini 的 functionCall 回合位置约束使其在重排前直接报 400，修复后合法且时序正确。
- 测试：`tests/test_openai_source.py` 更新为重排函数的直接调用（16 个用例）；`tests/test_anthropic_kimi_code_provider.py` 新增 2 个用例验证 Anthropic 格式转换后的时序；`tests/test_gemini_source.py` 新增 2 个用例验证 Gemini 的 messages 重排与转换后 `user → model(functionCall) → user(functionResponse)` 结构。

- [x] This is NOT a breaking change. / 这不是一个破坏性变更。


### Screenshots or Test Results / 运行截图或测试结果




```
$ python -m pytest tests/test_openai_source.py -k reorder -q
16 passed

$ python -m pytest tests/test_anthropic_kimi_code_provider.py -q
29 passed

$ python -m pytest tests/test_gemini_source.py -q
5 passed

$ python -m ruff check astrbot/core/provider/
All checks passed!
```

**Comment by Rail1bc** [comment]
@sourcery-ai review

**Comment by Rail1bc** [comment]
本次改动在框架层修复了伪造工具调用（fake tool call）消息对的时序错乱问题，覆盖 OpenAI / Anthropic / Gemini /
  Responses 四条 provider 路径：
  ▎ - 新增共享函数 reorder_tailing_tool_call_user /
  strip_internal_markers（astrbot/core/provider/provider.py），将插件注入的 assistant(tool_calls) → tool
  对重排到用户消息之后，与真实工具调用时序对齐；
  ▎ - 误伤防护：真实工具执行产生的对通过 _from_real_tool_call
  内部标记识别（经序列化/持久化/重试全链路保留），重排扫描遇到即停止；标记在发送边界对副本剥离，绝不进入发往 API
  的请求；
  ▎ - Gemini 路径修复后不再触发 400 INVALID_ARGUMENT 的 functionCall 位置约束报错；

  ▎ 测试：4 个相关测试文件 139 passed、3 failed（3 个失败为 Windows 平台预存在的 file_uri_to_path
  问题，与本次改动无关，上游基线同样失败）；ruff check 全部通过。
@xiaoyuyu6420 希望能抽空跟进此pr，谢谢！

**Comment by Rail1bc** [comment]
@sourcery-ai review

**Comment by Rail1bc** [comment]
@sourcery-ai review

**Comment by Rail1bc** [comment]
@sourcery-ai review

File: astrbot/core/provider/provider.py [hunk 2]
```diff
@@ -211,6 +217,96 @@ async def test(self, timeout: float = 45.0) -> None:
         )
 
 
+def _is_fake_tool_pair(asst_msg: dict[str, Any], tool_msg: dict[str, Any]) -> bool:
```
[the remark this answers is not in the document]
Comment (Rail1bc) [reply astrbot/core/provider/provider.py:220]: 1.已经改为直线检查序列
2.与之前的审查意见相悖。
3."发送边界"前提只对 OpenAI 成立

**Comment by Rail1bc** [comment]
@sourcery-ai review

File: astrbot/core/provider/provider.py [hunk 3]
```diff
@@ -211,6 +217,105 @@ async def test(self, timeout: float = 45.0) -> None:
...
```
[the remark this answers is not in the document]
Comment (Rail1bc) [reply astrbot/core/provider/provider.py:220]: Comment 2-1（拆分结构/语义谓词）—— 驳回

  行为完全等价（已逐分支核对）。但当前直线形式已在 docstring
  中分解为"先结构校验、后语义排除"两阶段，与代码一一对应；拆成 _is_tool_pair_candidate + _tool_pair_ids +
  _is_fake_tool_pair 三个单调用方 helper 属碎片化，且会反转第 3
  轮评审要求并已接受的合并。当前实现的线性校验链已满足可读性。

  Comment 2-2（切片式 block 收集）—— 驳回

  行为等价（单对/多对/标记阻断/非 dict/None 全部形态已手推验证）。但 (start, end) + 用 start == end 判空比当前的 (start,
   pairs) + not pairs 更隐晦，且丢失了对的实际数据；这是对核心收集逻辑的第三次改型，无行为收益的 churn
  只会增加回归风险。

  Comment 2-3（messages_to_payload 包装）—— 驳回

  该包装在目标调用点（发送边界）的两个分支是死代码：各 provider 的 context_query / payloads["messages"] 均为已序列化的
  dict（Message 对象在 _ensure_message_to_dicts 阶段已转换），checkpoint
  在上下文构建时已剥离（strip_checkpoint_messages）。它在真实调用点退化为普通
  strip_internal_markers，且实质重提了此前已评估过的注入/剥离合并方案。

**Comment by whatevertogo** [comment]
我结合 AstrBot 当前的上下文组装链路做了一个可运行原型，确认更适合复用现有的 `ProviderRequest.tool_calls_result`，而不是在 provider 发送前扫描消息并推断 fake/real。

具体做法是：

- 插件在 `on_llm_request` 中调用 `req.append_tool_calls_result(...)`
- `ToolLoopAgentRunner.reset()` 先组装当前 user 消息
- 然后追加 `request.tool_calls_result`

最终顺序天然为：

history
→ current user
→ assistant(tool_calls)
→ tool(result)

验证结果：

- 测试覆盖了完整的 `ToolLoopAgentRunner → OpenAI provider → SDK create()` 路径
- 测试中主动禁用了 `reorder_tailing_tool_call_user`
- 最终消息顺序仍然正确
- Runner 测试 40 个通过
- Checkpoint/集成测试 17 个通过
- Ruff 和格式检查通过
- 使用 `deepseek-v4-flash` 实网请求返回 200，并正确读取预计算工具结果

因此可以确认，这条路径不依赖启发式 reorder，也不依赖 `_from_real_tool_call` marker。

建议当前 PR 改为：

1. 由 runner 消费 `ProviderRequest.tool_calls_result`
2. 从 `astrbot.api.provider` 导出 `ToolCallsResult`、`AssistantMessageSegment` 和 `ToolCallMessageSegment`
3. 将相关插件迁移到 `req.append_tool_calls_result()`
4. 删除尾部 reorder、`_from_real_tool_call` marker，以及 marker 在序列化、持久化、modalities、retry 和各 provider 中的传播
5. 补充临时注入是否持久化的明确语义

这个方案让消息生产者显式表达注入位置，由 runner 统一负责顺序，provider 只负责格式转换，职责边界更清晰，也不会误伤旧历史或外部传入的真实工具调用。

**Comment by Rail1bc** [comment]
已经在新的pr里使用更好的修复方案

**Comment by xiaoyuyu6420** [comment]
感谢 @Rail1bc @whatevertogo 的深入分析和可运行原型！看了 #9673 的方案——由 `ToolLoopAgentRunner` 统一消费 `tool_calls_result` 通道，插件显式声明注入位置，顺序天然正确，而且 `_from_real_tool_call` 标记穿透序列化/持久化/重试的那整套复杂度都消失了。认同这个方向明显更优，本 PR 关闭没有异议。这次学到的重要一课：让消息生产者显式表达意图，优于在 provider 发送边界做启发式推断。
